jupyterhub_chameleon/handler.py

```diff
@@ -213,86 +216,40 @@ async def get(self):
-        )
 
-        response = None
-        trust_project_name = os.environ.get(
-            "ARTIFACT_SHARING_TRUST_PROJECT_NAME", "trovi"
+        # Authenticate to Trovi with our OpenStack credentials,
+        # exchanging our keystone token for a Trovi token
+        trovi_resp = requests.post(
+            urljoin(os.getenv("TROVI_URL"), "/token/"),
+            headers={"Content-Type": "application/json", "Accept": "application/json"},
+            json={
+                "grant_type": "token_exchange",
+                "subject_token": auth_state.get("access_token"),
+                "subject_token_type": "urn:ietf:params:oauth:token-type:jwt",
+                "scope": "artifacts:read artifacts:write",
+            },
         )
 
-        try:
-            trustee_user_id = user_session.get_user_id()
-            trustor_user_id = admin_session.get_user_id()
-            trust_project = next(
-                iter(admin_ks_client.projects.list(name=trust_project_name)), None
-            )
-
-            if not trust_project:
-                raise ValueError(
-                    (
-                        "Cannot create publish token because trust project "
-                        f"{trust_project_name} does not exist"
-                    )
-                )
+        trovi_token = trovi_resp.json()
 
-            trust = admin_ks_client.trusts.create(
-                trustee_user_id,
-                trustor_user_id,
-                project=trust_project,
-                role_names=["member"],
-            )
-            self.log.info(
-                (
-                    f"Created trust {trust.id} for user {self.current_user} "
-                    f"({trustee_user_id})"
-                )
-            )
+        self.log.debug(f"Trovi token exchange response: {trovi_token}")
 
-            trust_overrides = openstack_rc.copy()
-            trust_overrides["OS_TRUST_ID"] = trust.id
-            # Ensure no project-scoping keys are set on the keystone session;
-            # this will interfere with trust-scoping.
-            project_scoping_keys = [
-                "OS_PROJECT_NAME",
-                "OS_PROJECT_DOMAIN_NAME",
-                "OS_TENANT_NAME",
-                "OS_TENANT_DOMAIN_NAME",
-                "OS_PROJECT_ID",
-                "OS_TENANT_ID",
-            ]
-            for k in project_scoping_keys:
-                if k in trust_overrides:
-                    del trust_overrides[k]
-            trust_session = keystone_session(env_overrides=trust_overrides)
-
-            deposition_id = str(uuid.uuid4())
-            response = dict(
-                deposition_id=deposition_id,
-                publish_endpoint=dict(
-                    url=upload_url(deposition_id),
-                    method="PUT",
-                    headers=trust_session.get_auth_headers(),
-                ),
+        if trovi_resp.status_code != requests.codes.created:
```

Comment:
If the token endpoint throws a 404 or something it would be interpreted as unauthorized. Would it make sense to make this more narrow? In any event, I think we should log the response here, in particular if it's something other than what we expect from Trovi (401? 403?) -- otherwise difficult to debug any future problems.

Reply:
Good call.